### PR TITLE
[WIP] Port get-buffer-create

### PR DESCRIPTION
**Pull request by nitishch**
Ports `get-buffer-create`. I didn't port the functions `reset_buffer` and `reset_buffer_local_variables` as the branch became bigger than I expected.

Pending questions:

- I couldn't understand what `XSETBUFFER (buffer, b);` (line no. 463 in `buffer.c) does.
- Why are the `bool_bf` fields in `Lisp_Buffer` and `Lisp_Buffer_Text` combined to a single `flags` member? At least temporarily, I put the fields as they are in C code.

I guess because of the above, the code doesn't compile now.

Other changes:
- Introduced `Lisp_Interval` type
- Made `buffers::current_buffer()` return a `LispBufferRef` as opposed to a `LispObject`

### CODE REVIEW

**Comment by shaleh** [comment]
I suggest you wait for the bindgen PR to land. Then you will have a more stable place to work.

**Comment by shaleh** [comment]
From src/lisp.h
`#define XSETBUFFER(a, b) (XSETPSEUDOVECTOR (a, b, PVEC_BUFFER))`

Basically, many of the types in Emacs are internally stored as pseudo vectors. The various `XSET` macros mark the internals so the code knows what kind of vector it is.

File: rust_src/src/buffers.rs [hunk 1]
```diff
@@ -333,6 +378,105 @@ impl LispBufferRef {
         }
         self.local_flags[idx] = val;
     }
+
```
Comment (shaleh) [inline rust_src/src/buffers.rs]: Perhaps this should be a method on `LispBuffeRef` that allocates followed by something like this `create_new` which makes it ready for use?

**Comment by shaleh** [comment]
From the Rust perspective the bitfields are opaque blobs of memory. That is why they are a flags field. Checkout #781 for details.

**Comment by Wilfred** [comment]
bindgen has landed -- is there anything we can do to help you @nitishch?

**Comment by nitishch** [comment]
I will start fresh in a new PR.

**Comment by brotzeit** [comment]
I reopen this so your work will not be forgotten/lost. You can open a new PR or use this one again when you want to continue your work.

**Comment by shaleh** [comment]
Superseded by new PR

**Pull request closed without merging**